### PR TITLE
feat: build scripts (rust 1.81)

### PR DESCRIPTION
**Pull request by dj8yfo**
- **chore: add components to rust-toolchain.toml**
- **chore: update rust 1.75.0 -> 1.81.0 in rust-toolchain.toml**
- **chore: add better errors in tests around `std::fs:read` of wasm inputs**
- **chore: add `--no-docker` everywhere after `cargo near build`**
- **feat: add build script for `community` -> `discussions`**
- **feat: add build script for `community-factory` -> `community`**
- **chore: add building of contracts for tests into tests themselves**

### CODE REVIEW

**Comment by dj8yfo** [comment]
@race-of-sloths ping

**Comment by race-of-sloths** [comment]
@dj8yfo Thank you for your contribution! Your pull request is now a part of the [Race of Sloths](https://race-of-sloths.com)!
Weekly streak is on the road, smart strategy! Secure your streak with another PR!

[<picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://badge.race-of-sloths.com/dj8yfo?type=bot&pr=NEAR-DevHub/neardevhub-contract/149&theme=dark">
    <source media="(prefers-color-scheme: light)" srcset="https://badge.race-of-sloths.com/dj8yfo?type=bot&pr=NEAR-DevHub/neardevhub-contract/149&theme=light">
    Shows profile picture for the author of the PR
</picture>](https://race-of-sloths.com/profile/dj8yfo)





<details><summary>Current status: <i>executed</i></summary>

| Reviewer | Score |
|--------|--------|
| @petersalomonsen  | 5 |
| @frol  | 8 |


Your contribution is much appreciated with a final score of 6!
You have received 63 (60 base + 5% lifetime bonus) Sloth points for this contribution



</details><details><summary>What is the Race of Sloths</summary>
<p>
Race of Sloths is a friendly competition where you can participate in challenges and compete with other open-source contributors within your normal workflow

For contributors:

- Tag @race-of-sloths inside your pull requests
- Wait for the maintainer to review and score your pull request
- Check out your position in the [Leaderboard](https://race-of-sloths.com/leaderboard)
- Keep weekly and monthly streaks to reach higher positions
- Boast your contributions with a dynamic picture of your [Profile](https://race-of-sloths.com/profile/dj8yfo)

For maintainers:
- Score pull requests that participate in the Race of Sloths
- Engage contributors with fair scoring and fast responses so they keep their streaks
- Promote the Race to the point where the Race starts promoting you
- Grow the community of your contributors

Feel free to check [our website](https://race-of-sloths.com) for additional details!
</p>
</details> 

<details>
<summary>Bot commands</summary>

- For contributors
  - **Include a PR:** `@race-of-sloths include` to enter the Race with your PR
- For maintainers:
  - **Invite contributor** `@race-of-sloths invite` to invite the contributor to participate in a race or include it, if it's already a runner.
  - **Assign points:** `@race-of-sloths score [1/2/3/5/8/13]` to award points based on your assessment.
  - **Reject this PR:** `@race-of-sloths exclude` to send this PR back to the drawing board.
  - **Exclude repo:** `@race-of-sloths pause` to stop bot activity in this repo until `@race-of-sloths unpause` command is called

</details>

**Comment by petersalomonsen** [comment]
@frol @dj8yfo Should we merge https://github.com/NEAR-DevHub/neardevhub-contract/pull/148 first? It also has the `--no-docker` build setup.

File: .devcontainer/post-create.sh
Comment (petersalomonsen) [inline .devcontainer/post-create.sh]: Cargo-near fails to install in the ghcr.io/near/near-devcontainer image.

2024-09-10 19:39:47.137Z: downloading cargo-near 0.8.2 x86_64-unknown-linux-gnu
2024-09-10 19:39:48.739Z: installing to /usr/local/cargo/bin
2024-09-10 19:39:48.743Z: cp: cannot create regular file '/usr/local/cargo/bin/cargo-near': Permission denied
2024-09-10 19:39:48.748Z: ERROR(B[m: command failed: cp /tmp/tmp.NHfv1kPqDY/cargo-near /usr/local/cargo/bin

Consider applying the change from here:

https://github.com/NEAR-DevHub/neardevhub-contract/pull/148/files#diff-24ad71c8613ddcf6fd23818cb3bb477a1fb6d83af4550b0bad43099813088686

**Comment by petersalomonsen** [comment]
@race-of-sloths score 5

**Comment by petersalomonsen** [comment]
Great to see the new build scripts and that they are now integrated into the tests!

**Review by petersalomonsen — APPROVED** [review]
Looking good to me, but please check the codespace issue ( consider merging https://github.com/NEAR-DevHub/neardevhub-contract/pull/148 first )

File: .devcontainer/post-create.sh
Comment (dj8yfo) [reply .devcontainer/post-create.sh]: @petersalomonsen i overlooked that change in `.devcontainer/devcontainer.json` when copying from #148 ,
only copied from `.devcontainer/post-create.sh`. 

i'll add it too

**Comment by dj8yfo** [comment]
@petersalomonsen  i've added back `ghcr.io/near/near-devcontainer:latest` .
it appears to have everything needed 
https://github.com/NEAR-DevHub/neardevhub-contract/pull/149/commits/e5b5e448590cd1ea1b226dfe78470e14c88d5d62

**Comment by petersalomonsen** [comment]
> @petersalomonsen i've added back `ghcr.io/near/near-devcontainer:latest` . it appears to have everything needed [e5b5e44](https://github.com/NEAR-DevHub/neardevhub-contract/commit/e5b5e448590cd1ea1b226dfe78470e14c88d5d62)

Great ! Thanks!

File: community-factory/Cargo.toml
Comment (frol) [inline community-factory/Cargo.toml]: Is there a benefit in this style of defining the dependency over:

```toml
[build-dependencies]
cargo-near-build = { version = "0.1.0", features = ["build_script"] }
```

I find the standard way more common and readable/searchable

File: rust-toolchain.toml
Comment (frol) [inline rust-toolchain.toml:3]: I am not completely convinced if this line is needed at all. I hope it doesn't affect CI time, if so, I don't care much, just feels unnecessary

File: community/build.rs
Comment (frol) [inline community/build.rs]: Feels unnecessary for a one-time call of the macro. I'd just inline it, but it is not a blocker

File: README.md
Comment (frol) [inline README.md]: Wouldn't the contracts be automatically rebuilt based on the hints in build.rs?

File: tests/test_env.rs
Comment (frol) [inline tests/test_env.rs]: I'd build it regardless of whether the file exists to ensure that the test runs against the current version of the code (see my comment in the README). My hope is that the incremental build for no-changes will be instant.

**Review by frol — CHANGES_REQUESTED** [review]

File: rust-toolchain.toml
Comment (dj8yfo) [reply rust-toolchain.toml:3]: as discussed out of band, this helps installing rust-analyzer automatically with the right toolchain for anyone, 
 having rustup and checking out the repository.
 It does affect CI time, which must be around a second or two, so might be considered negligible

File: tests/test_env.rs
Comment (dj8yfo) [reply tests/test_env.rs]: hopefully handled by https://github.com/NEAR-DevHub/neardevhub-contract/pull/149/commits/6d44ceb17f69abb472fab5ba2b5eb89598c61d12

**Review by frol — APPROVED** [review]

**Comment by frol** [comment]
@race-of-sloths score 8

**Pull request merged**
